### ISSUE TITLE
Need to properly report rate limit responses (HTTP code 429)

### ISSUE THREAD

**Issue by nick-ivanov-ibm** [IBM-Cloud/sql-query-clients#59]
The client currently throws SyntaxError after receiving a rate limit response from the API, which is misleading and does not allow proper error handling. 

Consider raising a custom exception in such cases.

https://github.com/IBM-Cloud/sql-query-clients/blob/ce2453dedac765ff5a8ae8793b8fde99c5228323/Python/ibmcloudsql/SQLQuery.py#L148

**Comment by nick-ivanov-ibm** [comment]
It's a bit more convoluted than I thought. Here's the actual stack trace when 429 is issued:

```
2020-03-30T22:07:57.719520Z    stderr: Traceback (most recent call last):
2020-03-30T22:07:57.719525Z    stderr: File "/usr/local/lib/python3.7/site-packages/ibmcloudsql/SQLQuery.py", line 138, in submit_sql
2020-03-30T22:07:57.719529Z    stderr: body=json.dumps(sqlData))
2020-03-30T22:07:57.719533Z    stderr: File "/usr/local/lib/python3.7/site-packages/tornado/httpclient.py", line 102, in fetch
2020-03-30T22:07:57.719537Z    stderr: self._async_client.fetch, request, **kwargs))
2020-03-30T22:07:57.719541Z    stderr: File "/usr/local/lib/python3.7/site-packages/tornado/ioloop.py", line 458, in run_sync
2020-03-30T22:07:57.719545Z    stderr: return future_cell[0].result()
2020-03-30T22:07:57.719548Z    stderr: File "/usr/local/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
2020-03-30T22:07:57.719552Z    stderr: raise_exc_info(self._exc_info)
2020-03-30T22:07:57.719556Z    stderr: File "<string>", line 4, in raise_exc_info
2020-03-30T22:07:57.719559Z    stderr: tornado.httpclient.HTTPError: HTTP 429: Too Many Requests
2020-03-30T22:07:57.719563Z    stderr:
2020-03-30T22:07:57.719567Z    stderr: During handling of the above exception, another exception occurred:
2020-03-30T22:07:57.719571Z    stderr:
2020-03-30T22:07:57.719574Z    stderr: Traceback (most recent call last):
2020-03-30T22:07:57.719578Z    stderr: File "/action/1/src/main__.py", line 257, in sql_query_submit
2020-03-30T22:07:57.719582Z    stderr: job_id = resource.submit_sql(qry)
2020-03-30T22:07:57.719586Z    stderr: File "/usr/local/lib/python3.7/site-packages/ibmcloudsql/SQLQuery.py", line 140, in submit_sql
2020-03-30T22:07:57.719589Z    stderr: raise SyntaxError("SQL submission failed: {}".format(json_decode(e.response.body)['errors'][0]['message']))
2020-03-30T22:07:57.719593Z    stderr: File "<string>", line None
2020-03-30T22:07:57.719597Z    stderr: SyntaxError: SQL submission failed: This instance is currently running its maximum number of query jobs. Try again later, after at least one of the currently running jobs has completed.
```

**Comment by torsstei** [comment]
IMHO the submit_sql() should start retrying for a defined number of times when rate limit occurs. Maybe default 3 times. User should be able to change that default, e.g. also 0 to fail immediately.

**Comment by nick-ivanov-ibm** [comment]
May be, may be not. The point is, the client must be specifically informed about rate limit responses to take appropriate actions, which depends on a use case. You  don't want to retry immediately, because it's likely to result in an another 429. So you need a delay; but what delay is sensible? Only the client knows. And then you block the client while you're retrying, which might cause another failure (e.g. you're called from a cloud function and want to fail fast rather then block retrying, because the latter will cause cloud function timeout).

TLDR: I'd let the client decide if it wants to retry.

**Comment by dpittner** [comment]
Thanks, we'll likely push a enhancement for this today, we'll keep you in the loop

**Comment by nick-ivanov-ibm** [comment]
Thanks, you're too fast :) I was planning to submit a PR in a couple of days, time permitting.

**Comment by torsstei** [comment]
@nick-ivanov-ibm v.0.3.13 available on pypi.
